shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java

```diff
@@ -229,4 +229,27 @@ public static String failMetricName(String metricName) {
             return metricName + "_fail";
         }
     }
+
+    /**
+     * Generate a dot separated combination of metric name and tag value[s].
+     * e.g. metric "pravega.segmentstore.container.append_count" with tag "container=6" will get:
+     * "pravega.segmentstore.container.append_count.6".
+     *
+     * @param metric the metric name.
+     * @param tags the tag(s) associated with the metric.
+     * @return the dot separated combination of metric name and tag value(s).
+     */
+    public static String nameFromTags(String metric, String... tags) {
+        if (tags == null || tags.length == 0) {
+            return metric;
+        }
+        if (tags.length % 2 == 1) {
+            throw new IllegalArgumentException("tags size must be even, it is a set of key=value pairs");
+        }
+        StringBuilder sb = new StringBuilder(metric);
```

Comment:
From what I can tell below, this method is used extensively throughout the code, everytime we report something, and sometimes on the hot path. As such, we need to make this as efficient as possible. 

In this method you inefficiently use StringBuilder to append a number of different strings, then run the result through the extremely inefficient `escapeSpecialChar` method (just open it up to see what it does).

I recomment we make this O(n) with a single pass through.

Reply:
Micrometer docs make mention of how to use its API efficiently, in the bullet points here:
https://micrometer.io/docs/concepts#_global_registry

I think it is counter-productive to compose metric names on the fly.   The metric name may well be a constant; the tags provide the dynamic context such as `container_id`.  In other words, eliminate this method altogether.

@kevinhan88 mentioned that there's a back-compat concern, and I would suggest solving that in a different way, e.g. with a custom metric registry that encodes the metric name in the old way.

Reply:
@EronWright yes our existing implementation is very heavy to look up metric object always starting from forming search key string from scratch. 
However seems it's not possible to directly store strong reference to metric inside Pravega particularly for those "dynamic" (not global) metrics, such as segment level metrics, because Pravega should remain state-less in the sense, but low level metric objects are obviously state objects.
Having said that, the life-cycle management of metric objects can still be significantly simplified by taking advantage of Micrometer's weak referencing to metric objects.

@andreipaduroiu - yes the lookup of metric object is costly, however this refactoring doesn't introduce any additional burden to it:
With the old implementation, to look up metric with dimension, we have to call "nameFromTransaction()", "nameFromSegment", "nameFromContainer" or "nameFromHost" to form lookup key from scratch. (we do this to avoid direct strong reference to metric object, which is still valid). 
For example, to update counter for metric "append_counter" with container id 6, we call nameFromContainer("append_count", 6) to form the lookup key, which is "append_count.6", and pass this key to metric register to locate the metric object.
With the introduction of tags, we no longer call "nameFromContainer", instead we call nameFromTags("append_count", "containerId", "6") to form lookup key, which is "append_count.6". We still need this lookup key to manage lifecycle of this particular metric object (e.g. when container is closed, all the related metric objects should be evicted)
But when we call metric register, we do no use the lookup key above, we call register this way: register.search("append_count").tags("containerId", "6").counter()...

The point I'm trying to make, "nameFromTags()" is doing similar things as the old implementation does. There is a big space to improve this lookup mechanism, particularly considering the weak reference of Micrometer registry, however, it is in addition to the refactoring here, which is to replace Dropwizard with Micrometer to support dimensional data model. I'll create new issue to refactor the code further. Before that, we need to see Pravega works well with Telegraf, InfluxDB etc.

Reply:
My understanding of a dynamic logger is that it is a logger that includes instance data.   In the past, the dimensional data was packed into the metric name before the metric was looked up from the registry.  The reason for packing the data into the metric name is because codahale does not natively understand tags, and neither do many of the protocols.

Micrometer is different; it handles rewriting of the metric name as needed on a per-protocol basis.  For example, it transforms periods `.` to underscores `_` for some protocols.   For the statsd protocol, micrometer transforms the metric name in a manner similar to what `nameForX` does.   For the InfluxDB protocol, the metric name should not be transformed because InfluxDB protocol natively understands tags.  See [dimensionality](https://micrometer.io/docs/concepts#_supported_monitoring_systems).

I believe that the metric names may now be a constant, even when the metric is multi-dimensional.

Reply:
Agree, with the support of tag, the more general metric name should be a constant. 
I'm thinking to store tags into thread-local, or something similar. Hence lookup will be efficient as forming lookup key with metric name and tags is efficient, also the life-cycle mgmt is handled naturally with thread. No concrete solution yet.

Indeed it seems the teams are inspired by the Micrometer's programming model, makes it even more urgent to revisit our existing metric caching/lookup and life-cycle management mechanisms. Great!

Reply:
I see no need for a cache of metric objects.  Is it not as simple as:
```
import io.micrometer.Metrics;
Metrics.counter(
        Constants.PRAVEGA_SEGMENTSTORE_CONTAINER_APPENDCOUNT, 
        Constants.CONTAINER_ID, containerId).
    append();
```

Reply:
Yes got your point. We also need to manage the life-cycle of these metric objects. E.g the containers keep going up and down, end up with many objects inside register(s), but when should register clean them up? 
Micrometer register holds weak reference to these objects, so it depends on the strong reference from application. Seems Micrometer expects application to manage application objects life-cycle in some way, and metric objects just stay with application objects. If application object is gone, so does the strong reference to the metric object, then the metric objects is eligible for garbage collection, which marks the end of its life cycle.

Reply:
What Micrometer promotes is even further, it encourages application to store object directly, in the above case the APPEND_COUNTER_CONTAINER_6, and application directly updates this value, register directly export this value. No lookup at all end to end!
I'm sure we can come up with solid solution eventually.

Reply:
Why do you feel that Pravega need maintain a cache of meters?  My understanding is that the application need not keep references to the meters.  [Section 3.2](https://micrometer.io/docs/concepts#_global_registry) says:
> When tags need to be determined from local context, you have no choice but to construct/lookup the Meter inside your method body. The lookup cost is just a single hash lookup, so it will be acceptable for most uses.

Reply:
Metric registry must know when to terminate metric objects, otherwise it will run out of memory.
Micrometer terminates metric objects when:
1. all the strong reference from application ended;
2. registry.remove(metric) is explicitly called.

Pravega cache calls registry.remove() when time's up.
I'm not defending Pravega metric cache :-), I'm also thinking the way to get rid of it.

Reply:
@kevinhan88 Just because a window is broken, you don't need to replace it with yet another broken window.

Similarly, just because the old code was inefficient is not an excuse to write inefficient code. I pointed out something that use a lot of CPU for the value that it produces and which is used on the hot path. Everything that we use on the hot path adds to latency and decreases our performance. 

So please try to figure out a better way of redoing this code.

Reply:
@andreipaduroiu I created a new class MetricKey to hold both cache key and registry key, in order to make things backwards compatible, and as efficient as possible.